microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceDiagnosticsParticipant.java

```diff
@@ -119,26 +125,71 @@ private void validateMethod(MethodDeclaration node, List<Diagnostic> diagnostics
-										context.getUtils().toRange(openable, fallbackMethodExpr.getStartPosition(),
-												fallbackMethodExpr.getLength()),
-										DIAGNOSTIC_SOURCE, FALLBACK_METHOD_DOES_NOT_EXIST);
-								d.setSeverity(DiagnosticSeverity.Error);
-								diagnostics.add(d);
-							}
-						}
+						validateFallbackAnnotation(node, diagnostics, context, annotation);
+					} else if (isMatchAnnotation(annotation, ASYNCHRONOUS_ANNOTATION)) {
+						validateAsynchronousAnnotation(node, diagnostics, context, annotation);
 					}
 				}
 			}
 		}
+		
+		/**
+		 * Checks if the given method declaration has a fallback annotation, and if so,
+		 * provides diagnostics for the fallbackMethod
+		 *
+		 * @param node        The method declaration to validate
+		 * @param diagnostics A list where the diagnostics will be added
+		 * @param context     The context, used to create the diagnostics
+		 * @param annotation  The @Fallback annotation 
+		 * @throws JavaModelException
+		 */
+		private void validateFallbackAnnotation(MethodDeclaration node, List<Diagnostic> diagnostics,
+				JavaDiagnosticsContext context, NormalAnnotation annotation) throws JavaModelException {
+			Expression fallbackMethodExpr = getAnnotationMemberValueExpression(annotation,
+					FALLBACK_METHOD_FALLBACK_ANNOTATION_MEMBER);
+			if (fallbackMethodExpr != null) {
+				String fallbackMethodName = fallbackMethodExpr.toString();
+				fallbackMethodName = fallbackMethodName.substring(1, fallbackMethodName.length() - 1);
+				if (!getExistingMethods(node).contains(fallbackMethodName)) {
+					IOpenable openable = context.getTypeRoot().getOpenable();
+					Diagnostic d = context.createDiagnostic(context.getUri(),
+							"The referenced fallback method '" + fallbackMethodName + "' does not exist",
+							context.getUtils().toRange(openable, fallbackMethodExpr.getStartPosition(),
+									fallbackMethodExpr.getLength()),
+							DIAGNOSTIC_SOURCE, FALLBACK_METHOD_DOES_NOT_EXIST);
+					d.setSeverity(DiagnosticSeverity.Error);
+					diagnostics.add(d);
+				}
+			}
+		}
+		
+		/**
+		 * Checks if the given method declaration has an asynchronous annotation, and if so,
+		 * provides diagnostics for the method return type
+		 *
+		 * @param node        The method declaration to validate
+		 * @param diagnostics A list where the diagnostics will be added
+		 * @param context     The context, used to create the diagnostics
+		 * @param annotation  The @Asynchronous annotation 
+		 * @throws JavaModelException
+		 */
+		private void validateAsynchronousAnnotation(MethodDeclaration node, List<Diagnostic> diagnostics,
+				JavaDiagnosticsContext context, NormalAnnotation annotation) throws JavaModelException {
+			Type methodReturnType = node.getReturnType2();
+			String methodReturnTypeString = methodReturnType.resolveBinding().getQualifiedName();
+			LOGGER.log(Level.WARNING, "The method return type is " + methodReturnTypeString);
+			if (methodReturnType == null || 
```

Comment:
If the method return type is Future or CompletableFuture, then we would still emit diagnostics because of either the 2nd or 3rd conditions being true individually, due to the `||` .

Reply:
Ah sorry thats supposed to be &&

Reply:
Another quick tip. If you reverse the comparison, so "string".equals(someVar) then you don't need to check for null because the equals will handle it correctly.